syntax = "proto3";

package odin.v1;

enum ModuleStatus {
  INITIALIZING = 0;
  RUNNING      = 1;
  FAILED       = 2;
  DISCONNECTED = 3;
}

message Init {
  string session_id = 1;
  bytes  config     = 2;
}

message Shutdown {
  string session_id = 1;
}

message Command {
  oneof cmd {
    Init     init     = 1;
    Shutdown shutdown = 2;
  }
}

message Status {
  string       session_id = 1;
  string       module     = 2;
  ModuleStatus status     = 3;
  string       detail     = 4;   // error text if FAILED
}

// New messages for module bootup tracking
message ModuleBootup {
  string module_name = 1;
  int64  started_at  = 3;  // Unix timestamp
  string version     = 4;  // module version
  string host        = 5;  // hostname/IP where module is running
  string config_schema = 6; // JSON schema for module config
}

message ModuleHeartbeat {
  string module_name = 1;
  int64  timestamp   = 3;  // Unix timestamp of heartbeat
  ModuleStatus status = 4;
  int32  active_sessions = 5;  // number of active sessions
}

message ModuleRegistry {
  repeated ModuleBootup modules = 1;
  int64 last_updated = 2;
}
